File: app/src/main/java/com/itba/runningMate/rundetails/RunDetailsPresenter.java
package com.itba.runningMate.rundetails;

import android.graphics.Bitmap;
import android.net.Uri;

import com.itba.runningMate.domain.Route;
import com.itba.runningMate.domain.Run;
import com.itba.runningMate.repository.achievementsstorage.AchievementsStorage;
import com.itba.runningMate.repository.run.RunRepository;
import com.itba.runningMate.repository.runningstate.RunningStateStorage;
import com.itba.runningMate.rundetails.model.RunMetricsDetail;
import com.itba.runningMate.utils.providers.files.CacheFileProvider;
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider;

import java.io.File;
import java.io.FileOutputStream;
import java.lang.ref.WeakReference;

import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

public class RunDetailsPresenter {

    private final WeakReference<RunDetailsView> view;
    private final RunRepository runRepository;
    private final SchedulerProvider schedulerProvider;
    private final CacheFileProvider cacheFileProvider;
    private final RunningStateStorage runStorage;
    private final AchievementsStorage achievementsStorage;
    private final long runId;

    private double distance;
    private RunMetricsDetail detail;

    private final CompositeDisposable disposables;

    public RunDetailsPresenter(final CacheFileProvider cacheFileProvider,
                               final RunRepository runRepository,
                               final SchedulerProvider schedulerProvider,
                               final RunningStateStorage storage,
                               AchievementsStorage achievementsStorage, final long runId,
                               final RunDetailsView view) {
        this.view = new WeakReference<>(view);
        this.cacheFileProvider = cacheFileProvider;
        this.runRepository = runRepository;
        this.schedulerProvider = schedulerProvider;
        this.runStorage = storage;
        this.achievementsStorage = achievementsStorage;
        this.runId = runId;
        this.disposables = new CompositeDisposable();
    }

    public void onViewAttached() {
        disposables.add(runRepository.getRunMetrics(runId)
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onReceivedRunMetrics, this::onReceivedRunMetricsError));
    }

    private void onReceivedRunMetrics(Run run) {
        if (view.get() == null) {
            return;
        }
        distance = run.getDistance();
        detail = RunMetricsDetail.from(run);
        view.get().showRunMetrics(detail);
    }

    public void onMapAttached() {
        disposables.add(runRepository.getRun(runId)
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onReceivedRun, this::onReceivedRunMetricsError));
    }

    private void onReceivedRunMetricsError(Throwable throwable) {
        Timber.d("Failed to retrieve run route from db for run-id: %l", runId);
        if (view.get() != null) {
            view.get().showRunNotAvailableError();
        }
    }

    private void onReceivedRun(Run run) {
        if (view.get() != null) {
            view.get().showRoute(Route.from(run.getRoute()));
        }
    }

    public void onViewDetached() {
        disposables.dispose();
    }

    public void onDeleteButtonClick() {
        achievementsStorage.decreaseTotalDistance(this.distance);
        achievementsStorage.persistState();

        disposables.add(runRepository.deleteRun(runId)
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onRunDeleted, this::onRunDeleteError));
    }

    public void onRunTitleModified(String newTitle) {
        disposables.add(runRepository.updateTitle(runId, newTitle)
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.computation())
                .subscribe(this::onRunTitleUpdated, this::onRunTitleUpdateError));
    }

    public void onRunTitleUpdated() {
        Timber.i("Successfully updated title");
    }

    public void onRunTitleUpdateError(Throwable throwable) {
        Timber.d("Failed to update title");
        if (view.get() != null) {
            view.get().showUpdateTitleError();
        }
    }

    public void onShareButtonClick() {
        if (view.get() == null) {
            return;
        }
        Bitmap bitmap = view.get().getMetricsImage(detail);
        File image = cacheFileProvider.getFile("runningmate-run-metrics.png");
        Uri uri = null;
        try {
            FileOutputStream outputStream = new FileOutputStream(image);
            bitmap.compress(Bitmap.CompressFormat.PNG, 90, outputStream);
            outputStream.flush();
            outputStream.close();
            uri = cacheFileProvider.getUriForFile(image);
        } catch (Exception e) {
            view.get().showShareRunError();
        }
        view.get().shareImageIntent(uri);
    }

    private void onRunDeleted() {
        if (view.get() != null) {
            view.get().endActivity();
        }
    }

    private void onRunDeleteError(Throwable throwable) {
        Timber.d("Failed to delete run from db for run-id: %l", runId);
        if (view.get() != null) {
            view.get().showDeleteError();
        }
    }

}

File: app/src/main/java/com/itba/runningMate/rundetails/RunDetailsPresenter.kt
package com.itba.runningMate.rundetails

import com.itba.runningMate.domain.Route
import com.itba.runningMate.domain.Run
import com.itba.runningMate.repository.aggregaterunmetrics.AggregateRunMetricsStorage
import com.itba.runningMate.repository.run.RunRepository
import com.itba.runningMate.rundetails.model.RunMetricsDetail
import com.itba.runningMate.rundetails.model.RunMetricsDetail.Companion.from
import com.itba.runningMate.utils.ImageProcessing
import com.itba.runningMate.utils.providers.files.CacheFileProvider
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber
import java.io.FileOutputStream
import java.lang.ref.WeakReference

class RunDetailsPresenter(
    private val cacheFileProvider: CacheFileProvider,
    private val runRepository: RunRepository,
    private val schedulerProvider: SchedulerProvider,
    private val runId: Long,
    view: RunDetailsView?
) {
    private val view: WeakReference<RunDetailsView> = WeakReference(view)
    private val disposables: CompositeDisposable = CompositeDisposable()

    private var distance = 0.0
    private lateinit var detail: RunMetricsDetail

    fun onViewAttached() {
        fetchRun(runId)
    }

    private fun fetchRun(id: Long) {
        disposables.add(runRepository.getRunMetrics(id)
            .subscribeOn(schedulerProvider.computation())
            .observeOn(schedulerProvider.ui())
            .subscribe({ run: Run -> onReceivedRunMetrics(run) }) { onReceivedRunMetricsError() })
    }

    private fun onReceivedRunMetrics(run: Run) {
        if (view.get() == null) {
            return
        }
        distance = run.distance!!.toDouble()
        detail = from(run)
        view.get()!!.showRunMetrics(detail)
    }

    fun onMapAttached() {
        disposables.add(runRepository.getRun(runId)
            .subscribeOn(schedulerProvider.computation())
            .observeOn(schedulerProvider.ui())
            .subscribe({ run: Run -> onReceivedRun(run) }) { onReceivedRunMetricsError() })
    }

    private fun onReceivedRunMetricsError() {
        Timber.d("Failed to retrieve run route from db for run-id: %d", runId)
        if (view.get() != null) {
            view.get()!!.showRunNotAvailableError()
        }
    }

    private fun onReceivedRun(run: Run) {
        if (view.get() != null) {
            view.get()!!.showRoute(Route.from(run.route))
        }
    }

    fun onViewDetached() {
        disposables.dispose()
    }

    fun onDeleteButtonClick() {
        disposables.add(runRepository.deleteRun(runId)
            .subscribeOn(schedulerProvider.computation())
            .observeOn(schedulerProvider.ui())
            .subscribe({ onRunDeleted() }) { onRunDeleteError() })
    }

    fun onRunTitleModified(newTitle: String) {
        disposables.add(runRepository.updateTitle(runId, newTitle)
            .subscribeOn(schedulerProvider.computation())
            .observeOn(schedulerProvider.computation())
            .subscribe({ onRunTitleUpdated() }) { onRunTitleUpdateError() })
    }

    fun onRunTitleUpdated() {
        Timber.i("Successfully updated title")
    }

    private fun onRunTitleUpdateError() {
        Timber.d("Failed to update title")
        if (view.get() != null) {
            view.get()!!.showUpdateTitleError()
        }
    }

    fun onShareButtonClick() {
        if (view.get() == null) {
            return
        }
        val image = cacheFileProvider.getFile("runningmate-run-metrics.png")
        try {
            val outputStream = FileOutputStream(image)
            ImageProcessing.compress(view.get()!!.getMetricsImage(detail), outputStream)
        } catch (e: Exception) {
            view.get()!!.showShareRunError()
        }
        view.get()!!.shareImageIntent(cacheFileProvider.getUriForFile(image))
    }

    private fun onRunDeleted() {
        if (view.get() != null) {
            view.get()!!.endActivity()
        }
    }

    private fun onRunDeleteError() {
        Timber.d("Failed to delete run from db for run-id: %d", runId)
        if (view.get() != null) {
            view.get()!!.showDeleteError()
        }
    }

}